### DISCUSSION TITLE
Reordering annotation on clustermap

### DISCUSSION THREAD

**Discussion by sruthi-hub** [DingWB/PyComplexHeatmap#29] [Q&A] [answered]
First, thanks for making the complexheatmap package for python users. 
I am using the DotClustermapPlotter to make a dotplot with annotations for rows and columns. I want to specify the order for teh annotation of the columns. 
I see that class(AnnotationBase) has a [reorder](https://dingwb.github.io/PyComplexHeatmap/build/html/PyComplexHeatmap.annotations.html#PyComplexHeatmap.annotations.AnnotationBase.reorder) function. It would be nice to see an example to reorder the values in column annotation. 

For example ([provided here](https://dingwb.github.io/PyComplexHeatmap/build/html/notebooks/dotHeatmap.html#Visualize-up-to-five-dimension-data-using-DotClustermapPlotter)), if I want to change the order of the column such that I want Dataset 2 and then Dataset 1, how do I do that? Would appreciate your help. 

-----------------------
```
row_ha = HeatmapAnnotation(
                           Category=anno_simple(df_row.Category,cmap='Set1',
                                           add_text=False,legend=False),
                           label=anno_label(df_row.Category, merge=True,rotation=0),
                           axis=0,verbose=0,label_kws={'rotation':45,'horizontalalignment':'left'})

col_ha = HeatmapAnnotation(
                           Dataset=anno_simple(df_col.Dataset,cmap='Set1',legend=False,add_text=True),
                           Correlation=anno_simple(df_col.Correlation,cmap='Dark2',legend=False,add_text=True),
                           verbose=0,label_side='left',label_kws={'horizontalalignment':'right'})

plt.figure(figsize=(5, 8))
cm = DotClustermapPlotter(data=data, x='ID',y='Term',value='-log10(Pval)',c='-log10(Pval)',s='odds_ratio',
                          hue='EnrichType', row_cluster=False,col_cluster=False,
                          cmap={'Enrich':'RdYlGn_r','Depletion':'coolwarm_r'},
                          colors={'Enrich':'red','Depletion':'blue'},
                          #marker={'Enrich':'^','Depletion':'v'},
                          top_annotation=col_ha,right_annotation=row_ha,
                          col_split=df_col.Dataset,row_split=df_row.Category, col_split_gap=0.5,row_split_gap=1,
                          show_rownames=True,show_colnames=False,row_dendrogram=False,
                          verbose=0,legend_gap=7,alpha=0.8)
#plt.savefig("dotHeatmap1.pdf",bbox_inches='tight')
plt.show()
```
--------------------

**Comment by DingWB** [comment, marked as answer]
Please try `col_split=df_col.Dataset, col_split_order=['Dataset2','Dataset1']`. Or you can reorder the columns yourself and then specify `col_cluster=False`.

**Reply by sruthi-hub** [reply]
Thanks a lot, your suggestion worked! 
I also have some trouble displaying clustering rows and displaying dendrogram. 
When I do this:

```
row_ha = HeatmapAnnotation(
                           Category=anno_simple(df_row.Category,cmap='Set1',
                                           add_text=False,legend=False),
                           label=anno_label(df_row.Category, merge=True,rotation=0),
                           axis=0,verbose=0,label_kws={'rotation':45,'horizontalalignment':'left'})

col_ha = HeatmapAnnotation(
                           Dataset=anno_simple(df_col.Dataset,cmap='Set1',legend=False,add_text=True),
                           Correlation=anno_simple(df_col.Correlation,cmap='Dark2',legend=False,add_text=True),
                           verbose=0,label_side='left',label_kws={'horizontalalignment':'right'})

plt.figure(figsize=(5, 8))
cm = DotClustermapPlotter(data=data, x='ID',y='Term',value='-log10(Pval)',c='-log10(Pval)',s='odds_ratio',
                          hue='EnrichType', row_cluster=True,col_cluster=False,
                          cmap={'Enrich':'RdYlGn_r','Depletion':'coolwarm_r'},
                          colors={'Enrich':'red','Depletion':'blue'},
                          #marker={'Enrich':'^','Depletion':'v'},
                          top_annotation=col_ha,#right_annotation=row_ha,
                          col_split=df_col.Dataset,#row_split=df_row.Category, 
                          col_split_gap=0.5,#row_split_gap=1,
                          show_rownames=True,show_colnames=False,row_dendrogram=True,
                          verbose=0,legend_gap=7,alpha=0.8)
```

I get this error: 

`ValueError: The condensed distance matrix must contain only finite values.`

Is there something I am missing?

**Comment by DingWB** [comment]
There is an issue with your input data matrix. If two columns or rows are the same or finite values were included, then you would get such kind of error.
If you set `row_cluster=False,col_cluster=False`, there should be no error.

**Reply by sruthi-hub** [reply]
You're right, there is no error now. But I cannot display the row dendrogram if I do row_cluster=False, correct? 
Ideally I would like to display the dendrogram on one side, rownames on the other, and remove the row_ha (right_annotation). 
Is this possible?

**Comment by DingWB** [comment]
But if there was some problem with the data, then you could not perform clustering using scipy.

**Comment by sruthi-hub** [comment]
Not sure if I understand what you mean by problem with the data. The code above is from the [tutorial](https://dingwb.github.io/PyComplexHeatmap/build/html/notebooks/dotHeatmap.html#Visualize-up-to-five-dimension-data-using-DotClustermapPlotter) for the package here.
The data used in the tutorial is on [this package's github repo](https://github.com/DingWB/PyComplexHeatmap/blob/main/data/kycg_result.txt).  
I am facing similar problems with clustering even with my own data, so I think I may be using wrong parameters or there is some other issue.

**Comment by DingWB** [comment]
For this dataset (kycg_result.txt), it is too sparse and it could not be clustered. If you got an error, then the data can not be clustered.
Why don't you perform clustering directly using scipy, it you got an error even using Scipy, then it is the issue of the dataset.
